source-mysql/discovery.go

```diff
@@ -158,18 +180,107 @@ func getColumns(ctx context.Context, conn *client.Conn) ([]sqlcapture.ColumnInfo
 
 	var columns []sqlcapture.ColumnInfo
 	for _, row := range results.Values {
+		var typeName = string(row[5].AsString())
+		var dataType interface{}
+		if typeName == "enum" {
+			dataType = &mysqlColumnType{Type: "enum", EnumValues: parseEnumValues(string(row[6].AsString()))}
+		} else if typeName == "set" {
+			dataType = &mysqlColumnType{Type: "set", EnumValues: parseEnumValues(string(row[6].AsString()))}
+		} else {
+			dataType = typeName
+		}
 		columns = append(columns, sqlcapture.ColumnInfo{
 			TableSchema: string(row[0].AsString()),
 			TableName:   string(row[1].AsString()),
 			Index:       int(row[2].AsInt64()),
 			Name:        string(row[3].AsString()),
 			IsNullable:  string(row[4].AsString()) != "NO",
-			DataType:    string(row[5].AsString()),
+			DataType:    dataType,
 		})
 	}
 	return columns, err
 }
 
+type mysqlColumnType struct {
+	Type       string   `json:"type" mapstructure:"type"`           // The basic name of the column type.
+	EnumValues []string `json:"enum,omitempty" mapstructure:"enum"` // The list of values which an enum (or set) column can contain.
+}
+
+func (t *mysqlColumnType) translateRecordField(val interface{}) (interface{}, error) {
+	logrus.WithFields(logrus.Fields{
+		"type":  fmt.Sprintf("%#v", t),
+		"value": fmt.Sprintf("%#v", val),
+	}).Info("translating record field")
+
+	switch t.Type {
+	case "enum":
+		if index, ok := val.(int64); ok {
+			if 1 <= index && index <= int64(len(t.EnumValues)) {
+				return t.EnumValues[index-1], nil
+			}
+		} else if bs, ok := val.([]byte); ok {
+			return string(bs), nil
+		}
+		return val, nil
+	case "set":
+		if bitfield, ok := val.(int64); ok {
+			var acc strings.Builder
+			for idx := 0; idx < len(t.EnumValues); idx++ {
+				if bitfield&(1<<idx) != 0 {
+					if acc.Len() > 0 {
+						acc.WriteByte(',')
+					}
+					acc.WriteString(t.EnumValues[idx])
+				}
+			}
+			return acc.String(), nil
+		} else if bs, ok := val.([]byte); ok {
+			return string(bs), nil
+		}
+		return val, nil
+	}
+	return val, fmt.Errorf("error translating value of complex column type %q", t.Type)
+}
+
+var enumValuesRegexp = regexp.MustCompile(`'((?:''|\\.|[^'])+)'(?:,|$)`)
```

Comment:
I'm just trusting you on this one ;)

Reply:
Yeah, sorry, it's one of those regexes that looks really complicated because it's made entirely of line noise, but I'm decently confident it's correct.

TL;DR: it matches a single-quoted string followed by either a comma or EOL, using non-capturing groups for the alternations on the terminator and the string body characters. The options for string body characters are, in order: "two successive quotes", "anything backslash-escaped", and "any character that's not a single-quote".